### ISSUE TITLE
No error emitted for `use clippy;`

### ISSUE THREAD

**Issue by camelid** [rust-lang/rust#83334]
```rust
use clippy::a;
```

errors with

```
error[E0432]: unresolved import `clippy`
 --> src/lib.rs:1:5
  |
1 | use clippy::a;
  |     ^^^^^^ `clippy` is a tool module, not a module
```

but

```rust
use clippy;
```

just says

```
warning: unused import: `clippy`
 --> src/lib.rs:1:5
  |
1 | use clippy;
  |     ^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
```

Note that actually *using* the import gives an error:

```
error: cannot use a tool module through an import
 --> src/main.rs:4:5
  |
4 |     clippy::a();
  |     ^^^^^^
  |
note: the tool module imported here
 --> src/main.rs:1:5
  |
1 | use clippy;
  |     ^^^^^^

error[E0423]: expected function, found tool attribute `clippy::a`
 --> src/main.rs:4:5
  |
4 |     clippy::a();
  |     ^^^^^^^^^ not a function

error: aborting due to 2 previous errors
```

Found this while playing around with the MCVE in #83317.

**Comment by petrochenkov** [comment]
This is working as expected, `use clippy;` working is just uniform import paths in action.
Any name in scope can be imported, including tool modules.

"Cannot use through an import" errors are reported for attributes that are detected textually by their users (tool attributes by tools, derive helpers by proc macros, built-in attributes by rustc), and renaming would break their use.

`use clippy::a;` should arguably work in theory, but it's hard to do due to implementation details of imports (the import path is split to two parts, the first one of which must be a module), and also entirely useless.

**Comment by camelid** [comment]
Hmm, interesting. By the way, why doesn't `use clippy as foo;` work?

**Comment by jyn514** [comment]
Interestingly, this doesn't work with `register_tool` either (use 'clippy' on the right, the warning is denied by default): https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=252d78a905b7b628c04f8315542bd150

**Comment by petrochenkov** [comment]
>By the way, why doesn't `use clippy as foo; #[deny(foo::warnings)] ...` work?

Because lint naming has zero relation to name resolution.
`foo::warnings` is not a path in a regular sense, it's not resolved in the current scope, it's just some path-like syntax to name lints belonging to tools.

**Comment by camelid** [comment]
> > By the way, why doesn't `use clippy as foo; #[deny(foo::warnings)] ...` work?
> 
> Because lint naming has zero relation to name resolution.
> `foo::warnings` is not a path in a regular sense, it's not resolved in the current scope, it's just some path-like syntax to name lints belonging to tools.

I was asking because I thought tool modules might be allowed to be imported because you could rename them, but I guess it's niche enough that there hasn't been a need. I'll close this issue then.
